aten/src/THNN/generic/SpatialGridSamplerBilinear.c

```diff
@@ -56,59 +64,72 @@ TH_API void THNN_(SpatialGridSamplerBilinear_updateOutput)(
   for (n = 0; n < N; ++n) {
     for (h = 0; h < H; ++h) {
       for (w = 0; w < W; ++w) {
-	// get the corresponding input x, y co-ordinates from grid
-	real ix = THTensor_fastGet4d(grid, n, h, w, 0);
-	real iy = THTensor_fastGet4d(grid, n, h, w, 1);
-
-	// normalize ix, iy from [-1, 1] to [0, IH-1] & [0, IW-1]
-	ix = ((ix + 1) / 2) * (IW-1);
-	iy = ((iy + 1) / 2) * (IH-1);
-
-	// get NE, NW, SE, SW pixel values from (x, y)
-	int ix_nw = floor(ix);
-	int iy_nw = floor(iy);
-	int ix_ne = ix_nw + 1;
-	int iy_ne = iy_nw;
-	int ix_sw = ix_nw;
-	int iy_sw = iy_nw + 1;
-	int ix_se = ix_nw + 1;
-	int iy_se = iy_nw + 1;
-
-	// get surfaces to each neighbor:
-	real nw = (ix_se - ix)    * (iy_se - iy);
-	real ne = (ix    - ix_sw) * (iy_sw - iy);
-	real sw = (ix_ne - ix)    * (iy    - iy_ne);
-	real se = (ix    - ix_nw) * (iy    - iy_nw);
-	  
-	// calculate bilinear weighted pixel value and set output pixel
-	for (c = 0; c < C; ++c) {
-	  //   (c, iy_nw, ix_nw) * nw + (c, iy_ne, ix_ne) * ne
-	  // + (c, iy_sw, ix_sw) * sw + (c, iy_se, ix_se) * se
-	  real nw_val = SAFE_GET(input, ix_nw, iy_nw, n, c, IH, IW);
-	  real ne_val = SAFE_GET(input, ix_ne, iy_ne, n, c, IH, IW);
-	  real sw_val = SAFE_GET(input, ix_sw, iy_sw, n, c, IH, IW);
-	  real se_val = SAFE_GET(input, ix_se, iy_se, n, c, IH, IW);
-	  real out_val = nw_val * nw + ne_val * ne + sw_val * sw + se_val * se;
-	  THTensor_fastSet4d(output, n, c, h, w, out_val);
-	}
+        // get the corresponding input x, y co-ordinates from grid
+        real ix = THTensor_fastGet4d(grid, n, h, w, 0);
+        real iy = THTensor_fastGet4d(grid, n, h, w, 1);
+
+        // normalize ix, iy from [-1, 1] to [0, IH-1] & [0, IW-1]
+        ix = ((ix + 1) / 2) * (IW-1);
+        iy = ((iy + 1) / 2) * (IH-1);
+
+        // get NE, NW, SE, SW pixel values from (x, y)
+        int ix_nw = floor(ix);
+        int iy_nw = floor(iy);
+        int ix_ne = ix_nw + 1;
+        int iy_ne = iy_nw;
+        int ix_sw = ix_nw;
+        int iy_sw = iy_nw + 1;
+        int ix_se = ix_nw + 1;
+        int iy_se = iy_nw + 1;
+
+        // get surfaces to each neighbor:
+        real nw = (ix_se - ix)    * (iy_se - iy);
+        real ne = (ix    - ix_sw) * (iy_sw - iy);
+        real sw = (ix_ne - ix)    * (iy    - iy_ne);
+        real se = (ix    - ix_nw) * (iy    - iy_nw);
+
+        if (padding_mode==MODE_BORDER){
+          // clip coordinates to image borders
+          CLIP_COORDINATES(ix_nw, ix_nw, IW);
+          CLIP_COORDINATES(iy_nw, iy_nw, IH);
+          CLIP_COORDINATES(ix_ne, ix_ne, IW);
+          CLIP_COORDINATES(iy_ne, iy_ne, IH);
+          CLIP_COORDINATES(ix_sw, ix_sw, IW);
+          CLIP_COORDINATES(iy_sw, iy_sw, IH);
+          CLIP_COORDINATES(ix_se, ix_se, IW);
+          CLIP_COORDINATES(iy_se, iy_se, IH);
+        }
+
+        // calculate bilinear weighted pixel value and set output pixel
+        for (c = 0; c < C; ++c) {
+          //   (c, iy_nw, ix_nw) * nw + (c, iy_ne, ix_ne) * ne
```

Comment:
Can we remove the commented out bits of code

Reply:
Ah realize this is from the original version, just shows up as new as you switched tabs -> spaces.

Reply:
ideally one commit is dedicated to switching tabs to spaces, and the rest of the commits are rebased on top of that. otherwise code review becomes difficult.

Reply:
Apologies for this. I will take it into account for the future.

Tabs/spaces changes are limited to the files:
`aten/src/THNN/generic/SpatialGridSamplerBilinear.c`
`aten/src/THNN/generic/THNN.h`